zebra-chain/src/primitives/address.rs

```diff
@@ -93,7 +96,61 @@ impl zcash_address::TryFromAddress for Address {
             .ok_or_else(|| BoxError::from("not a valid sapling address").into())
     }
 
-    // TODO: Add sprout and unified/orchard converters
+    fn try_from_unified(
+        network: zcash_address::Network,
+        unified_address: zcash_address::unified::Address,
+    ) -> Result<Self, zcash_address::ConversionError<Self::Error>> {
+        let network = network.try_into()?;
+        let mut orchard = None;
+        let mut sapling = None;
+        let mut transparent = None;
+
+        for receiver in unified_address.items().into_iter() {
+            match receiver {
+                unified::Receiver::Orchard(data) => {
+                    orchard = orchard::Address::from_raw_address_bytes(&data).into();
+                    // ZIP 316: Consumers MUST reject Unified Addresses/Viewing Keys in
+                    // which any constituent Item does not meet the validation
+                    // requirements of its encoding.
+                    if orchard.is_none() {
+                        return Err(BoxError::from(
+                            "Unified Address contains an invalid Orchard receiver.",
+                        )
+                        .into());
+                    }
+                }
+                unified::Receiver::Sapling(data) => {
+                    sapling = sapling::PaymentAddress::from_bytes(&data);
+                    // ZIP 316: Consumers MUST reject Unified Addresses/Viewing Keys in
+                    // which any constituent Item does not meet the validation
+                    // requirements of its encoding.
+                    if sapling.is_none() {
+                        return Err(BoxError::from(
+                            "Unified Address contains an invalid Sapling receiver",
+                        )
+                        .into());
+                    }
+                }
+                unified::Receiver::P2pkh(data) => {
+                    transparent = Some(transparent::Address::from_pub_key_hash(network, data));
+                }
+                unified::Receiver::P2sh(data) => {
+                    transparent = Some(transparent::Address::from_script_hash(network, data));
+                }
+                unified::Receiver::Unknown { .. } => {
+                    return Err(BoxError::from("Unsupported receiver in a Unified Address.").into());
```

Comment:
I'm not sure if unknown receivers are invalid?

Unified addresses are designed to be forwards-compatible, so new or experimental receiver types can be added at any time. In this case, we can ignore them because the RPC doesn't have fields for them.